### PR TITLE
backport: Merge bitcoin#19237,19378,18422

### PR DESCRIPTION
**Pull request by vijaydasmp**

### CODE REVIEW

File: src/test/fuzz/fuzz.cpp [hunk 1]
```diff
@@ -43,6 +43,15 @@ void initialize()
...
 
+// Decide if main(...) should be provided:
+// * AFL needs main(...) regardless of platform.
+// * macOS handles __attribute__((weak)) main(...) poorly when linking
+//   against libFuzzer. See https://github.com/bitcoin/bitcoin/pull/18008
+//   for details.
+#if defined(__AFL_COMPILER) || !defined(MAC_OSX)
+#define PROVIDE_MAIN_FUNCTION
+#endif
+
```
Comment (UdjinM6) [inline src/test/fuzz/fuzz.cpp:54]: 19366: backported earlier via 50d20fe788 already, should be dropped

File: test/functional/p2p_invalid_messages.py [hunk 2]
```diff
@@ -9,12 +9,14 @@
 from test_framework.messages import (
     msg_ping,
     ser_string,
+    MAX_PROTOCOL_MESSAGE_LENGTH
```
Comment (UdjinM6) [inline test/functional/p2p_invalid_messages.py:12]: 19272: feels like we miss some previous backports... might want to investigate further/drop this one for now

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -274,6 +274,35 @@ int FindAndDelete(CScript& script, const CScript& b)
...
+ *
+ * A return value of false means the script fails entirely. When true is returned, the
+ * fSuccess variable indicates whether the signature check itself succeeded.
+ */
+static bool EvalChecksig(const valtype& vchSig, const valtype& vchPubKey, CScript::const_iterator pbegincodehash, CScript::const_iterator pend, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* serror, bool& fSuccess)
+{
+    // Subset of script starting at the most recent codeseparator
+    CScript scriptCode(pbegincodehash, pend);
+
+    // Drop the signature in pre-segwit scripts but not segwit scripts
```
Comment (UdjinM6) [inline src/script/interpreter.cpp:287]: 18422:
```suggestion
    // Drop the signature, since there's no way for a signature to sign itself
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/script/interpreter.cpp (hunk 3 above)
Comment (UdjinM6) [reply src/script/interpreter.cpp:287]: still not fixed, see https://github.com/dashpay/dash/pull/5143/commits/c88389400d9677454573f061311ab4a0ecea2f5b#diff-a0337ffd7259e8c7c9a7786d6dbd420c80abfa1afdb34ebae3261109d9ae3c19L984

Comment (vijaydasmp) [reply src/script/interpreter.cpp]: fixed now

**Review by UdjinM6 — APPROVED** [review]
utACK

**Comment by vijaydasmp** [comment]
Hello @PastaPastaPasta , please review

**Review by knst — DISMISSED** [review]
utACK

**Comment by UdjinM6** [comment]
@PastaPastaPasta

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**